exec/java-exec/src/main/java/org/apache/drill/exec/server/Drillbit.java

```diff
@@ -371,65 +376,91 @@ private void javaPropertiesToSystemOptions() {
-      final Path drillHomePath;
+      String drillHome = System.getenv(DRILL_HOME);
+      String gracefulFile = System.getenv(GRACEFUL_SIGFILE);
+      Path drillHomePath;
       if (drillHome == null || gracefulFile == null) {
-        logger.warn("Cannot access graceful file. Graceful shutdown from command line will not be supported.");
+        if (logger.isWarnEnabled()) {
+          StringBuilder builder = new StringBuilder(NOT_SUPPORTED_MESSAGE);
+          if (drillHome == null) {
+            builder.append(" ").append(DRILL_HOME).append(" is not set.");
+          }
+          if (gracefulFile == null) {
+            builder.append(" ").append(GRACEFUL_SIGFILE).append(" is not set.");
+          }
+          logger.warn(builder.toString());
+        }
         return;
       }
       try {
         drillHomePath = Paths.get(drillHome);
+        if (!Files.exists(drillHomePath)) {
+          logger.warn("{} path [{}] does not exist. {}", DRILL_HOME, drillHomePath, NOT_SUPPORTED_MESSAGE);
+          return;
+        }
       } catch (InvalidPathException e) {
-        logger.warn("Cannot access graceful file. Graceful shutdown from command line will not be supported.");
+        logger.warn("Unable to construct {}} path [{}]: {}. {}",
+          DRILL_HOME, drillHome, e.getMessage(), NOT_SUPPORTED_MESSAGE);
         return;
       }
-      boolean triggered_shutdown = false;
-      WatchKey wk = null;
-      try (final WatchService watchService = drillHomePath.getFileSystem().newWatchService()) {
-        drillHomePath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY, StandardWatchEventKinds.ENTRY_CREATE);
-        while (!triggered_shutdown) {
-          wk = watchService.take();
-          for (WatchEvent<?> event : wk.pollEvents()) {
-            final Path changed = (Path) event.context();
-            if (changed != null && changed.endsWith(gracefulFile)) {
-              drillbit.interruptPollShutdown = false;
-              triggered_shutdown = true;
-              drillbit.close();
-              break;
+
+      try (WatchService watchService = drillHomePath.getFileSystem().newWatchService()) {
+        drillHomePath.register(watchService, StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_MODIFY);
+        while (true) {
+          WatchKey watchKey = watchService.take();
+          for (WatchEvent<?> event : watchKey.pollEvents()) {
+            if (StandardWatchEventKinds.OVERFLOW != event.kind()) {
+              Path changedPath = (Path) event.context();
+              if (changedPath != null && changedPath.endsWith(gracefulFile)) {
+                drillbit.interruptPollShutdown = false;
+                drillbit.close();
+                return;
+              }
             }
           }
-        }
-      } finally {
-        if (wk != null) {
-          wk.cancel();
+          if (!watchKey.reset()) {
+            drillbit.interruptPollShutdown = false;
+            return;
+          }
```

Comment:
Thanks for adding the call to reset. But I think there is still some more changes needed here. If `watchkey.reset()` returns false, in our case that means we have not seen any event related to graceful file and still observing some failure while resetting the watchKey. In this case we should try to register again because if we returned from here that means GracefulShutdown thread will exit with Drillbit still up and running. Any future attempt for graceful shutdown will not work.

Reply:
Done.